packages/statemanager/src/ethersStateManager.ts

```diff
@@ -0,0 +1,400 @@
+    const codeBuffer = toBuffer(code)
+    this.contractCache.set(address.toString(), codeBuffer)
+    return codeBuffer
+  }
+
+  /**
+   * Adds `value` to the state trie as code, and sets `codeHash` on the account
+   * corresponding to `address` to reference this.
+   * @param address - Address of the `account` to add the `code` for
+   * @param value - The value of the `code`
+   */
+  async putContractCode(address: Address, value: Buffer): Promise<void> {
+    // Store contract code in the cache
+    this.contractCache.set(address.toString(), value)
+  }
+
+  /**
+   * Gets the storage value associated with the provided `address` and `key`. This method returns
+   * the shortest representation of the stored value.
+   * @param address -  Address of the account to get the storage for
+   * @param key - Key in the account's storage to get the value for. Must be 32 bytes long.
+   * @returns {Buffer} - The storage value for the account
+   * corresponding to the provided address at the provided key.
+   * If this does not exist an empty `Buffer` is returned.
+   */
+  async getContractStorage(address: Address, key: Buffer): Promise<Buffer> {
+    // Retrieve storage slot from provider if not found in cache
+    await this.getContractStorageFromProvider(address, key)
+
+    const storageTrie = await this._getStorageTrie(address)
+    const foundValue = await storageTrie.get(key) //@ts-ignore
+
+    return Buffer.from(RLP.decode(Uint8Array.from(foundValue ?? [])) as Uint8Array)
+  }
+
+  /**
+   * Retrieves a storage slot from the provider and stores in the local trie
+   * @param address Address to be retrieved from provider
+   * @param key Key of storage slot to be returned
+   * @private
+   */
+  private async getContractStorageFromProvider(address: Address, key: Buffer): Promise<void> {
+    if (this.externallyRetrievedStorageKeys.has(address.toString())) {
+      const map = this.externallyRetrievedStorageKeys.get(address.toString())
+      if (map?.get(key.toString('hex')) !== undefined) {
+        return
+      }
+    }
+
+    const accountData = await this.provider.send('eth_getProof', [
+      address.toString(),
+      [bufferToHex(key)],
+      this.blockTag,
+    ])
+    const rawData = accountData.accountProof
+    await this.trie.fromProof(rawData.map((e: string) => hexToBytes(e)))
+    // Only requesting a single slot at a time so the proof will always be the first item in the array
+    const storageData = accountData.storageProof[0]
+
+    const storageTrie = await this._getStorageTrie(address)
+
+    await storageTrie.fromProof(storageData.proof.map((e: string) => hexToBytes(e)))
+
+    let map = this.externallyRetrievedStorageKeys.get(address.toString())
+    if (!map) {
+      this.externallyRetrievedStorageKeys.set(address.toString(), new Map())
+    }
+    map = this.externallyRetrievedStorageKeys.get(address.toString())!
+    map.set(key.toString('hex'), true)
```

Comment:
```suggestion
    let map = this.externallyRetrievedStorageKeys.get(address.toString())
    if (!map) {
      map = this.externallyRetrievedStorageKeys
        .set(address.toString(), new Map())
        .get(address.toString())!
    }
    map.set(key.toString('hex'), true)
```
I think this works and avoids `get`ting twice when map is retrieved from the first `get`

Reply:
Actually this is even simpler
```suggestion
     const map =
      this.externallyRetrievedStorageKeys.get(address.toString()) ??
      this.externallyRetrievedStorageKeys
        .set(address.toString(), new Map())
        .get(address.toString())!
    map.set(key.toString('hex'), true)
```

Reply:
Hmm, at the cost of one additional boolean, I'd prefer just moving the second get inside the `!map` check.  I know we can chain all that stuff together but it just makes it less legible to me and I don't think any material difference in code executed, right?

Reply:
Yes, sure! No difference indeed, let's go with what makes it most legible